### PR TITLE
SONAR-8338 add properties to control log level of JMX related loggers

### PR DESCRIPTION
**Pull request by sns-seb**
based on #1395

### CODE REVIEW

File: server/sonar-process/src/main/java/org/sonar/process/LogbackHelper.java [hunk 1]
```diff
@@ -254,18 +260,61 @@ public boolean isAllLogsToConsoleEnabled(Props props) {
...
+    if (!isAllowed(level)) {
+      throw new IllegalArgumentException(String.format("log level %s in property %s is not a supported value (allowed levels are %s)",
+        level, propertyKey, Arrays.toString(ALLOWED_ROOT_LOG_LEVELS)));
+    }
+    return level;
+  }
+
+  private static boolean isAllowed(Level level) {
+    for (Level allowedRootLogLevel : ALLOWED_ROOT_LOG_LEVELS) {
+      if (level == allowedRootLogLevel) {
```
Comment (SonarTech) [inline server/sonar-process/src/main/java/org/sonar/process/LogbackHelper.java]: MAJOR Change this comparison to use the equals method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1698)

File: server/sonar-process/src/main/java/org/sonar/process/LogbackHelper.java [hunk 2]
```diff
@@ -275,13 +324,31 @@ public Level configureRootLogLevel(Props props, String propertyKey) {
...
+
+  public Level configureLoggerLogLevelFromDomain(String loggerName, Props props, ProcessId processId, LogDomain domain) {
+    return configureLoggersLogLevelFromDomain(Collections.singleton(loggerName), props, processId, domain);
+  }
+
+  public Level configureLoggersLogLevelFromDomain(Set<String> loggerNames, Props props, ProcessId processId, LogDomain domain) {
+    String processProperty = SONAR_PROCESS_LOG_LEVEL_PROPERTY.replace(PROCESS_NAME_PLACEHOLDER, processId.getKey());
+    Level newLevel = resolveLevel(props, SONAR_LOG_LEVEL_PROPERTY, processProperty, processProperty + "." + domain.key);
+    loggerNames.forEach(loggerName -> {
+      Logger logger = getRootContext().getLogger(loggerName);
```
Comment (SonarTech) [inline server/sonar-process/src/main/java/org/sonar/process/LogbackHelper.java]: MAJOR Make the "logger" logger private static final and rename it to comply with the format "LOG(?:GER)?". [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1312)

File: server/sonar-process/src/main/java/org/sonar/process/LogbackHelper.java [hunk 3]
```diff
@@ -254,18 +260,61 @@ public boolean isAllLogsToConsoleEnabled(Props props) {
...
-  public Level configureRootLogLevel(Props props, String propertyKey) {
-    Level newLevel = Level.toLevel(props.value(propertyKey, Level.INFO.toString()), Level.INFO);
-    return configureRootLogLevel(newLevel);
+  public Level configureRootLogLevel(Props props, String... propertyKeys) {
+    Level newLevel = resolveLevel(props, propertyKeys);
+    getRootContext().getLogger(ROOT_LOGGER_NAME).setLevel(newLevel);
+    return newLevel;
+  }
+
+  private Level resolveLevel(Props props, String... propertyKeys) {
```
Comment (SonarTech) [inline server/sonar-process/src/main/java/org/sonar/process/LogbackHelper.java]: MAJOR Make "resolveLevel" a "static" method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

File: sonar-db/src/main/java/org/sonar/db/DefaultDatabase.java [hunk 4]
```diff
@@ -93,7 +97,7 @@ private void initDataSource() throws Exception {
     datasource = new ProfiledDataSource(basicDataSource, NullConnectionInterceptor.INSTANCE);
     datasource.setConnectionInitSqls(dialect.getConnectionInitStatements());
     datasource.setValidationQuery(dialect.getValidationQuery());
-    enableSqlLogging(datasource, "TRACE".equals(settings.getString("sonar.log.level")));
+    enableSqlLogging(datasource, logbackHelper.getLoggerLevel("sql") == Level.TRACE);
```
Comment (SonarTech) [inline sonar-db/src/main/java/org/sonar/db/DefaultDatabase.java]: MAJOR Change this comparison to use the equals method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1698)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues
* MAJOR 4 major

Watch the comments in this conversation to review them.

File: server/sonar-process/src/main/java/org/sonar/process/Lifecycle.java [hunk 5]
```diff
@@ -101,4 +110,11 @@ public boolean equals(@Nullable Object o) {
   public int hashCode() {
     return state.hashCode();
   }
+
+  public interface LifecycleListener {
```
Comment (SonarTech) [inline server/sonar-process/src/main/java/org/sonar/process/Lifecycle.java]: CRITICAL Annotate the "LifecycleListener" interface with the @FunctionalInterface annotation [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1609)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues
* CRITICAL 1 critical
* MAJOR 4 major

Watch the comments in this conversation to review them.

File: sonar-application/src/main/assembly/conf/sonar.properties [hunk 6]
```diff
@@ -217,9 +217,60 @@
 #--------------------------------------------------------------------------------------------------
 # LOGGING
 
-# Level of logs. Supported values are INFO(default), DEBUG and TRACE (DEBUG + SQL + ES requests)
+# SonarQube produces logs in 4 logs files located in the same directory (see property sonar.path.logs below),
+# one per JVM: App (main process), Web Server (aka. Web), Compute Engine (aka. CE) and Elastic Search (aka. ES)
```
Comment (simonbrandhof) [inline sonar-application/src/main/assembly/conf/sonar.properties]: I suggest to replace "JVM" by "process". Moreover correct spelling of ES is "Elasticsearch".

File: sonar-application/src/main/assembly/conf/sonar.properties [hunk 7]
```diff
@@ -217,9 +217,60 @@
...
+#
+# 1: timestamp. Format is YYYY.MM.DD HH:MM:SS
+#    YYYY: year on 4 digits
+#    MM: month on 2 digits
+#    DD: day on 2 digits
+#    HH: hour of day on 2 digits in 24 hours format
+#    MM: minutes on 2 digits
+#    SS: seconds on 2 digits
+# 2: log level. Possible values are (in order of descending criticality): ERROR, WARN, INFO, DEBUG and TRACE
+# 3: process identifier. Possible values are: app (main process), web (Web Server), ce (Compute Engine) and es (Elastic Search)
```
Comment (simonbrandhof) [inline sonar-application/src/main/assembly/conf/sonar.properties]: "Elasticsearch"

File: sonar-application/src/main/assembly/conf/sonar.properties [hunk 8]
```diff
@@ -217,9 +217,60 @@
...
+# 4: thread identifier. Can be empty.
```
Comment (simonbrandhof) [inline sonar-application/src/main/assembly/conf/sonar.properties]: It's not exactly the thread identifier, for example startup logs don't have one. I suggest to replace with "context identifier" or "request identifier".

File: sonar-application/src/main/assembly/conf/sonar.properties [hunk 9]
```diff
@@ -217,9 +217,60 @@
...
+#    In the Web Server, if present, it will be the HTTP request identifier.
+#    In the Compute Engine, if present, it will be the task UUID.
```
Comment (simonbrandhof) [inline sonar-application/src/main/assembly/conf/sonar.properties]: not "UUID" but "identifier" or "ID".

File: sonar-application/src/main/assembly/conf/sonar.properties [hunk 10]
```diff
@@ -217,9 +217,60 @@
...
+# When specified, they overwrite the level defined at global level.
+# Supported values are INFO, DEBUG and TRACE
+#sonar.log.level.app=INFO
+#sonar.log.level.web=INFO
+#sonar.log.level.ce=INFO
+#sonar.log.level.es=INFO
+
+# Sql logs in the Web and CE processes can be individually controlled with their respective properties
+# Supported values are INFO, DEBUG and TRACE
+# Currently, there is Sql logs only at TRACE level
```
Comment (simonbrandhof) [inline sonar-application/src/main/assembly/conf/sonar.properties]: "there are"

File: sonar-application/src/main/assembly/conf/sonar.properties [hunk 11]
```diff
@@ -217,9 +217,60 @@
...
+#sonar.log.level.web.sql=TRACE
```
Comment (simonbrandhof) [inline sonar-application/src/main/assembly/conf/sonar.properties]: commented-out values are supposed to be the default values.
So here it should be INFO.

File: sonar-application/src/main/assembly/conf/sonar.properties [hunk 12]
```diff
@@ -217,9 +217,60 @@
...
+#sonar.log.level.ce.sql=TRACE
+
+# Es client logs in the Web and CE processes can be individually controlled with their respective properties
```
Comment (simonbrandhof) [inline sonar-application/src/main/assembly/conf/sonar.properties]: "Elasticsearch"

File: sonar-application/src/main/assembly/conf/sonar.properties (hunk 11 above)
Comment (sns-seb) [inline sonar-application/src/main/assembly/conf/sonar.properties]: there is no default value on this property.

or at least it's not a constant. It's actually the value of `sonar.log.level.web` if set or at least the one of `sonar.log.level`.

so, I decided to provide the one useful value  for this property in the commented code

File: sonar-application/src/main/assembly/conf/sonar.properties (hunk 6 above)
Comment (sns-seb) [inline sonar-application/src/main/assembly/conf/sonar.properties]: well... SQ has actually up to 5 processes : on windows the wrapper runs as a 5th process.
But SQ never has more than 4 JVMs.

File: sonar-application/src/main/assembly/conf/sonar.properties (hunk 8 above)
Comment (simonbrandhof) [inline sonar-application/src/main/assembly/conf/sonar.properties]: as a user I want to quickly understand which need it covers. "Functional thread" won't help as I don't know what it means.

File: sonar-application/src/main/assembly/conf/sonar.properties (hunk 11 above)
Comment (simonbrandhof) [inline sonar-application/src/main/assembly/conf/sonar.properties]: This inconsistency is misleading IMO. @christophelevis WDYT ?

File: sonar-application/src/main/assembly/conf/sonar.properties (hunk 6 above)
Comment (simonbrandhof) [inline sonar-application/src/main/assembly/conf/sonar.properties]: as a user I don't care of this technical detail. The term "process" is used everywhere in this file and in WS.

File: sonar-application/src/main/assembly/conf/sonar.properties (hunk 11 above)
Comment (sns-seb) [inline sonar-application/src/main/assembly/conf/sonar.properties]: which inconsistency? in the value of the commented property or the absence of default value?

Comment (simonbrandhof) [inline sonar-application/src/main/assembly/conf/sonar.properties]: commented property. It's usually the effective value (and indeed not necesarily the default value)

Comment (sns-seb) [inline sonar-application/src/main/assembly/conf/sonar.properties]: the effective value is the value of sonar.log.level or sonar.log.level.web, which can be any of INFO, DEBUG and TRACE

File: sonar-application/src/main/assembly/conf/sonar.properties (hunk 8 above)
Comment (sns-seb) [inline sonar-application/src/main/assembly/conf/sonar.properties]: I'll go for "SQ Thread identifier" or something of the like as it makes sense to reference the Thread concept. As so, fields go from larger to most precise in a sense that any one can understand : process -> thread -> logger

**Comment by sns-seb** [comment]
BTW, I've added the name of the log file in the sonar.properties

**Pull request merged**